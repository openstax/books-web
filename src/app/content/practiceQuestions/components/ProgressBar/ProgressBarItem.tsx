import React from 'react';
import styled, { css } from 'styled-components/macro';
import theme from '../../../../theme';

// tslint:disable-next-line: variable-name
const StyledItemWrapper = styled.span`
  display: flex;
  align-items: center;
  margin-bottom: 1rem;

  &::after {
    content: "";
    display: block;
    height: 0.1rem;
    width: 5.6rem;
    background-color: ${theme.color.primary.gray.base};
    ${(props: { isActiveOrDisabled: boolean }) => {
      if (props.isActiveOrDisabled) {
        return 'background-color: #F1F1F1;';
      }
    }}
  }

  &:last-child {
    &::after {
      content: none;
    }
  }

  ${theme.breakpoints.mobile(css`
    &::after {
      width: 3.3rem;
    }
  `)}
`;

// tslint:disable-next-line: variable-name
const StyledItem = styled.span`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 2.4rem;
  height: 2.4rem;
  border-radius: 50%;
  background-color: ${theme.color.primary.gray.base};
  border: 1px solid ${theme.color.primary.gray.base};
  color: ${theme.color.neutral.base};
  margin: 0 0.4rem;
  font-weight: bold;
  ${(props: { isActive: boolean, isDisabled: boolean }) => {
    if (props.isActive) {
      return `
        color: ${theme.color.primary.gray.base};
        background-color: ${theme.color.neutral.base};
      `;
    } else if (props.isDisabled) {
      return `
        color: #C1C1C1;
        background-color: #F1F1F1;
        border-color: #F1F1F1;
      `;
    }
  }}
`;

interface ProgressBarItemProps {
  value: number;
  isActive: boolean;
  isDisabled: boolean;
}

// tslint:disable-next-line: variable-name
const ProgressBarItem = ({ value, isActive, isDisabled }: ProgressBarItemProps) => (
  <StyledItemWrapper isActiveOrDisabled={isDisabled || isActive}>
    <StyledItem isActive={isActive} isDisabled={isDisabled}>
      {value}
    </StyledItem>
  </StyledItemWrapper>
);

export default ProgressBarItem;
