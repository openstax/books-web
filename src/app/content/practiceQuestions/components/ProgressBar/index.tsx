import React from 'react';
import styled from 'styled-components/macro';
import ProgressBarItem from './ProgressBarItem';

// tslint:disable-next-line: variable-name
const ProgressBarWrapper = styled.div`
  display: flex;
  justify-content: flex-start;
  flex-wrap: wrap;
`;

interface ProgressBarProps {
  total: number;
  activeIndex: number;
}

// tslint:disable-next-line: variable-name
const ProgressBar = ({ total, activeIndex }: ProgressBarProps) => {
  // total prop could be changed to questions: Question[] and then we wouldn't have to do this hack
  // with populating empty array
  const items: Array<() => JSX.Element> = new Array(total);

  for (let index = 0; index < total; index++) {
    items[index] = () => <ProgressBarItem
      key={index}
      value={index + 1}
      isActive={index === activeIndex}
      isDisabled={index > activeIndex}
    />;
  }

  return <ProgressBarWrapper>
    {items.map((item) => item())}
  </ProgressBarWrapper>;
};

export default ProgressBar;
