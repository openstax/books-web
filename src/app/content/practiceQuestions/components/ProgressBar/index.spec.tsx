import React from 'react';
import renderer from 'react-test-renderer';
import ProgressBar from '.';
import ProgressBarItem from './ProgressBarItem';

describe('ProgressBar', () => {
  it('displays properly', () => {
    const component = renderer.create(<ProgressBar total={3} activeIndex={1} />);

    const [item1, item2, item3] = component.root.findAllByType(ProgressBarItem);

    expect(item1.props.isActive).toEqual(false);
    expect(item1.props.isDisabled).toEqual(false);

    expect(item2.props.isActive).toEqual(true);
    expect(item2.props.isDisabled).toEqual(false);

    expect(item3.props.isActive).toEqual(false);
    expect(item3.props.isDisabled).toEqual(true);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
