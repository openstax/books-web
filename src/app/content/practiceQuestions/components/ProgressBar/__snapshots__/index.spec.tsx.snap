// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar displays properly 1`] = `
<div
  className="ProgressBar__ProgressBarWrapper-sc-12ssvtu-0 hVMttS"
>
  <span
    className="ProgressBarItem__StyledItemWrapper-sc-1d02vnm-0 IySic"
  >
    <span
      className="ProgressBarItem__StyledItem-sc-1d02vnm-1 bipRWq"
    >
      1
    </span>
  </span>
  <span
    className="ProgressBarItem__StyledItemWrapper-sc-1d02vnm-0 hnxurH"
  >
    <span
      className="ProgressBarItem__StyledItem-sc-1d02vnm-1 ProgressBarItem__StyledActiveItem-sc-1d02vnm-2 cAorVC"
    >
      2
    </span>
  </span>
  <span
    className="ProgressBarItem__StyledItemWrapper-sc-1d02vnm-0 hnxurH"
  >
    <span
      className="ProgressBarItem__StyledItem-sc-1d02vnm-1 ProgressBarItem__StyledDisabledItem-sc-1d02vnm-3 izehWs"
    >
      3
    </span>
  </span>
</div>
`;
